### PR TITLE
First draft of the docs for TF + Datasets

### PR DESCRIPTION
**Pull request by Rocketknight1**
I might cc a few of the other TF people to take a look when this is closer to being finished, but it's still a draft for now.

### CODE REVIEW

File: docs/source/use_with_tensorflow.mdx [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+or external storage. The dataset can be transformed arbitrarily with the `map()` method, or methods like `batch()`
+and `shuffle()` can be used to create a dataset that's ready for training. These methods do not modify the stored data
+in any way - instead, the methods build a data pipeline graph that will be executed when the dataset is iterated over,
+usually during model training or inference. This is different from the `map()` method of Hugging Face `Dataset` objects,
+which runs the map function immediately and saves the new or changed columns.
+
+Since the entire data preprocessing pipeline can be compiled in a `tf.data.Dataset`, this approach allows for massively
+parallel, asynchronous data loading and training. However, the requirement for graph compilation can be a limitation,
+particularly for HuggingFace tokenizers, which are usually not (yet!) compilable as part of a TF graph. As a result, for NLP
+tasks we usually advise pre-processing the dataset as a Hugging Face dataset, where arbitrary Python functions can be
```
Comment (lhoestq) [inline docs/source/use_with_tensorflow.mdx:43]: no need to mention "NLP", we can keep it general I think ;)
```suggestion
particularly for HuggingFace tokenizers, which are usually not (yet!) compilable as part of a TF graph. As a result, 
we usually advise pre-processing the dataset as a Hugging Face dataset, where arbitrary Python functions can be
```

File: docs/source/use_with_tensorflow.mdx [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+suffice, but for more complex tasks a custom collator may be necessary. In particular, many NLP tasks have samples
+with varying sequence lengths which will require a collator that can pad batches correctly. (Link to transformers
+collators or examples here?)
+
+### Caveats and limitations
+
+Right now, `to_tf_dataset()` focuses on making "ready-to-use" datasets with batching and data collation applied.
+However, particularly for computer vision tasks, we realize that it would often be preferable to convert a
+Hugging Face `Dataset` to an unbatched, unmodified `tf.data.Dataset` and then perform batching/shuffling using `tf.data`
+methods. We're actively working on adding this feature and making it performant, so watch this space!
```
Comment (lhoestq) [inline docs/source/use_with_tensorflow.mdx:77]:
```suggestion
Right now, `to_tf_dataset()` always return a batched dataset, we will add support for unbatched datasets soon !
```

File: docs/source/use_with_tensorflow.mdx [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+```
+
+The returned `tf_ds` object here is now fully ready to train on, and can be passed directly to `model.fit()`!
+
+For a full description of the arguments, please see the `to_tf_dataset()` documentation(link). In many cases,
+you will also need to add a `collate_fn` to your call. This is a function that takes multiple elements of the dataset
+and combines them into a single batch. When all elements have the same length, the built-in default collator will
+suffice, but for more complex tasks a custom collator may be necessary. In particular, many NLP tasks have samples
+with varying sequence lengths which will require a collator that can pad batches correctly. (Link to transformers
+collators or examples here?)
```
Comment (lhoestq) [inline docs/source/use_with_tensorflow.mdx:70]: Same here, maybe it can be more general than just NLP

**Review by lhoestq — COMMENTED** [review]
Thanks ! I haven't checked it all yet but here are my first comments.

I think this section can also be a good place to explain different use cases of `to_tf_dataset`

File: docs/source/use_with_tensorflow.mdx (hunk 3 above)
Comment (Rocketknight1) [reply docs/source/use_with_tensorflow.mdx:70]: Fixed!

**Comment by lhoestq** [comment]
Some links are still missing I think :)

File: docs/source/use_with_tensorflow.mdx [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+>>> text = ["foo", "bar"]
+>>> data = [0, 1] 
+>>> ds = Dataset.from_dict({"text": text, "data": data})  
+>>> ds = ds.with_format("tf") 
+>>> ds[:2]
+{'text': <tf.Tensor: shape=(2,), dtype=string, numpy=array([b'foo', b'bar'], dtype=object)>,
+ 'data': <tf.Tensor: shape=(2,), dtype=int64, numpy=array([0, 1])>}
+```
+
+But if you want, you can explicitly format certain columns and leave the other columns unformatted:
```
Comment (stevhliu) [inline docs/source/use_with_tensorflow.mdx:103]: Maybe write a sentence or two about scenarios in which you might want to do this?

```suggestion
You can also explicitly format certain columns and leave the other columns unformatted:
```

File: docs/source/use_with_tensorflow.mdx [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+The `tf.data.Dataset` class covers a wide range of use-cases - it is often created from Tensors in memory, or using a load function to read files on disc
+or external storage. The dataset can be transformed arbitrarily with the `map()` method, or methods like `batch()`
+and `shuffle()` can be used to create a dataset that's ready for training. These methods do not modify the stored data
+in any way - instead, the methods build a data pipeline graph that will be executed when the dataset is iterated over,
+usually during model training or inference. This is different from the `map()` method of Hugging Face `Dataset` objects,
+which runs the map function immediately and saves the new or changed columns.
+
+Since the entire data preprocessing pipeline can be compiled in a `tf.data.Dataset`, this approach allows for massively
+parallel, asynchronous data loading and training. However, the requirement for graph compilation can be a limitation,
+particularly for HuggingFace tokenizers, which are usually not (yet!) compilable as part of a TF graph. As a result,
```
Comment (stevhliu) [inline docs/source/use_with_tensorflow.mdx:133]:
```suggestion
particularly for Hugging Face tokenizers, which are usually not (yet!) compilable as part of a TF graph. As a result, 
```

File: docs/source/use_with_tensorflow.mdx [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+>>> ds[:2]
+{'data': <tf.Tensor: shape=(2,), dtype=int64, numpy=array([0, 1])>,
+ 'text': ['foo', 'bar']}
+```
+
+The [`Image`] and [`Audio`] feature types are not supported yet.
+
+## Data loading
+
+### Interoperability with tf.data
```
Comment (stevhliu) [inline docs/source/use_with_tensorflow.mdx:116]: We can probably remove this section header and let this content fall under the `## Data loading` header.

File: docs/source/use_with_tensorflow.mdx [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+in any way - instead, the methods build a data pipeline graph that will be executed when the dataset is iterated over,
+usually during model training or inference. This is different from the `map()` method of Hugging Face `Dataset` objects,
+which runs the map function immediately and saves the new or changed columns.
+
+Since the entire data preprocessing pipeline can be compiled in a `tf.data.Dataset`, this approach allows for massively
+parallel, asynchronous data loading and training. However, the requirement for graph compilation can be a limitation,
+particularly for HuggingFace tokenizers, which are usually not (yet!) compilable as part of a TF graph. As a result, 
+we usually advise pre-processing the dataset as a Hugging Face dataset, where arbitrary Python functions can be
+used, and then converting to `tf.data.Dataset` afterwards using `to_tf_dataset()` to get a batched dataset ready for
+training. To see examples of this approach, please see the example scripts(link) or notebooks(link) for `transformers`.
```
Comment (stevhliu) [inline docs/source/use_with_tensorflow.mdx]:
```suggestion
training. To see examples of this approach, please see the [example scripts](https://github.com/huggingface/transformers/tree/main/examples/tensorflow) or [notebooks](https://github.com/huggingface/transformers/tree/main/notebooks#tensorflow-examples) for `transformers`.
```

File: docs/source/use_with_tensorflow.mdx [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+```
+
+The returned `tf_ds` object here is now fully ready to train on, and can be passed directly to `model.fit()`! Note
+that you set the batch size when creating the dataset, and so you don't need to specify it when calling `fit()`:
+
+```py
+>>> model.fit(tf_ds, epochs=2)
+```
+
+For a full description of the arguments, please see the `to_tf_dataset()` documentation(link). In many cases,
```
Comment (stevhliu) [inline docs/source/use_with_tensorflow.mdx]:
```suggestion
For a full description of the arguments, please see the [`~Dataset.to_tf_dataset`] documentation. In many cases,
```

File: docs/source/use_with_tensorflow.mdx [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+you will also need to add a `collate_fn` to your call. This is a function that takes multiple elements of the dataset
+and combines them into a single batch. When all elements have the same length, the built-in default collator will
+suffice, but for more complex tasks a custom collator may be necessary. In particular, many tasks have samples
+with varying sequence lengths which will require a collator that can pad batches correctly. (Link to transformers
```
Comment (stevhliu) [inline docs/source/use_with_tensorflow.mdx]:
```suggestion
with varying sequence lengths which will require a [collator](https://huggingface.co/docs/transformers/main/en/main_classes/data_collator) that can pad batches correctly. (Link to transformers
```

**Review by stevhliu — APPROVED** [review]
Great explanation about how the data loading works and when to use it!

File: docs/source/use_with_tensorflow.mdx (hunk 6 above)
Comment (Rocketknight1) [reply docs/source/use_with_tensorflow.mdx:116]: Done! (Will push in a sec)

File: docs/source/use_with_tensorflow.mdx (hunk 7 above)
Comment (Rocketknight1) [reply docs/source/use_with_tensorflow.mdx]: I pushed this in my own branch, but it's fixed!

File: docs/source/use_with_tensorflow.mdx (hunk 8 above)
Comment (Rocketknight1) [reply docs/source/use_with_tensorflow.mdx]: Done!

File: docs/source/use_with_tensorflow.mdx [hunk 10]
```diff
@@ -0,0 +1,195 @@
+# Using Datasets with TensorFlow
+
+## Tensors
```
[the remark this answers is not in the document]
Comment (Rocketknight1) [reply docs/source/use_with_tensorflow.mdx]: Added!

File: docs/source/use_with_tensorflow.mdx [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+samples from it. If you wish to convert the entire dataset to `Tensor`, simply query the full dataset:
+
+```py
+>>> ds[:]
+{'data': <tf.Tensor: shape=(2, 2), dtype=int64, numpy=
+array([[1, 2],
+       [3, 4]])>}
+```
+
+### N-dimensional arrays
```
Comment (lhoestq) [inline docs/source/use_with_tensorflow.mdx:42]:
```suggestion
## N-dimensional arrays
```

**Review by lhoestq — APPROVED** [review]
Awesome thanks :)

**Comment by Rocketknight1** [comment]
This is probably quite close to being ready, so cc some TF people @gante @amyeroberts @merveenoyan just so they see it! No need for a full review, but if you have any comments or suggestions feel free.

**Pull request merged**